neo4j-cypher-dsl-parser/src/main/java/org/neo4j/cypherdsl/parser/Options.java

```diff
@@ -0,0 +1,210 @@
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.cypherdsl.parser;
+
+import static org.apiguardian.api.API.Status.EXPERIMENTAL;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+import org.apiguardian.api.API;
+import org.jetbrains.annotations.Nullable;
+import org.neo4j.cypherdsl.core.Clauses;
+import org.neo4j.cypherdsl.core.Expression;
+import org.neo4j.cypherdsl.core.Return;
+import org.neo4j.cypherdsl.core.utils.Assertions;
+
+/**
+ * Provides arguments to the {@link CypherParser cypher parser}. The options itself are thread safe and can be reused.
+ * The listener and modifications you provide are of course out of our control.
+ *
+ * @author Michael J. Simons
+ * @soundtrack Queen - The Game
+ * @since 2021.3.0
+ */
+@API(status = EXPERIMENTAL, since = "2021.3.0")
+public final class Options {
+
+	private static final Options DEFAULT_OPTIONS = newOptions().build();
+
+	/**
+	 * @return the default options
+	 */
+	public static Options defaultOptions() {
+		return DEFAULT_OPTIONS;
+	}
+
+	/**
+	 * Use this method to start defining new options for a parser.
+	 *
+	 * @return A builder for new options.
+	 */
+	public static Builder newOptions() {
+		return Builder.newConfig();
+	}
+
+	/**
+	 * Use this builder to create a new set of options..
+	 */
+	@SuppressWarnings("HiddenField")
+	public static final class Builder {
+
+		private BiFunction<LabelParsedEventType, Collection<String>, Collection<String>> labelFilter = (e, l) -> l;
+
+		private BiFunction<TypeParsedEventType, Collection<String>, Collection<String>> typeFilter = (e, t) -> t;
+
+		private final Map<ExpressionCreatedEventType, List<Function<Expression, ? extends Expression>>> onNewExpressionCallbacks = new HashMap<>();
+
+		private Function<ReturnDefinition, Return> returnClauseFactory;
+
+		private Builder() {
+		}
+
+		static Builder newConfig() {
+			return new Builder();
+		}
+
+		/**
+		 * Configure a filter that is applied to labels.
+		 *
+		 * @param labelFilter Takes in an event type, a collection of labels and returns a probably new collection of labels.
+		 * @return This builder.
+		 */
+		public Builder withLabelFilter(BiFunction<LabelParsedEventType, Collection<String>, Collection<String>> labelFilter) {
+
+			Assertions.notNull(labelFilter, "Label filter may not be null.");
+			this.labelFilter = labelFilter;
+			return this;
+		}
+
+		/**
+		 * Configure a filter that is applied to types.
+		 *
+		 * @param typeFilter Takes in an event type and the parsed type. May return the type itself or something else.
+		 * @return This builder.
+		 */
+		public Builder withTypeFilter(BiFunction<TypeParsedEventType, Collection<String>, Collection<String>> typeFilter) {
```

Comment:
Just to write this out: The `Collection<String>` as a return type of the function is intentionally provided, right?
We want to enable something like `[:Blubb]` -> `[:Bla|Wurstsalat]`

Reply:
Absolutely. Adding / removing one or more labels.